### PR TITLE
List and read Views

### PR DESCRIPTION
**Pull request by cristianvasquez**
I've added the following functionalities:

- To fetch a list of views from a source
- To fetch a view using the URI
- To build a view from {term, dataset}

I'm marking this as a draft since I haven't added enough tests.

### CODE REVIEW

File: lib/Dimension.js [hunk 1]
```diff
@@ -13,24 +13,27 @@ class Dimension extends Node {
...
-
-        if (join) {
-          from.addOut(ns.view.join, toTerm(join))
-        }
-      })
-      .addOut(ns.view.as, as || path)
-
-    if (aggregate) {
-      this.ptr.addOut(ns.view.aggregate, aggregate)
+    // If term is given, use it as a pointer, otherwise create the quads.
```
Comment (tpluscode) [inline lib/Dimension.js:16]: You say "otherwise" but there is no `else`...

File: lib/Filter.js [hunk 2]
```diff
@@ -11,19 +11,22 @@ class Filter extends Node {
       graph
     })
 
-    this.ptr
-      .addOut(ns.rdf.type, ns.view.Filter)
-      .addOut(ns.view.dimension, toTerm(dimension))
-      .addOut(ns.view.operation, operation)
+    // If term is given, use it as a pointer, otherwise create the quads.
```
Comment (tpluscode) [inline lib/Filter.js:14]: Ditto

File: lib/Node.js [hunk 3]
```diff
@@ -27,12 +27,24 @@ class Node {
...
 
     this.ptr.deleteOut()
   }
+
+  _detach () {
+    for (const child of this.parent.children) {
+      if (child.term.equals(this.ptr.term)) {
+        this.parent.children.delete(child)
+      }
+    }
```
Comment (tpluscode) [inline lib/Node.js:46]: You sure this needs to loop?

**Review by tpluscode — CHANGES_REQUESTED** [review]
I don't understand everything but I left some comments :)

File: lib/Filter.js [hunk 4]
```diff
@@ -44,7 +47,11 @@ class Filter extends Node {
   }
 
   clear () {
-    this.ptr.deleteList(ns.view.argument)
+    if (this.ptr.isList()) {
+      this.ptr.deleteList(ns.view.argument)
+    } else {
+      this.ptr.deleteOut(ns.view.argument)
+    }
```
[the remark this answers is not in the document]
Comment (bergos) [reply lib/Filter.js:54]: No, the update looks good. Arguments can be plain terms or lists. No need to make the structure unnecessary complex. See the [example](https://github.com/zazuko/rdf-cube-view-schema/blob/master/view.ttl#L122-L135).

**Review by bergos — CHANGES_REQUESTED** [review]
Good comments from @tpluscode which should be resolved before merging.

**Comment by cristianvasquez** [comment]
I will address the comments, thanks for the review!

File: lib/Node.js (hunk 3 above)
Comment (cristianvasquez) [reply lib/Node.js:46]: Sorry, I don't get the question.

One would like to do simply this.parent.children.delete(child), but the term only can be checked for equality using equals(), not ==

Or do you mean to move it into a functional style?

Comment (tpluscode) [reply lib/Node.js:46]: Maybe there was some other intent but the code above looks equivalent to

```js
this.parent.children.delete(this)
```

File: lib/View.js [hunk 5]
```diff
@@ -41,17 +44,28 @@ class View extends Node {
...
-      join,
-      path,
-      source
+  static fromCube (cube) {
+    function getCubeSource (cube) {
+      cube.ptr.addOut(ns.rdf.type, ns.view.CubeSource).addOut(ns.view.cube, toTerm(cube))
+      return cube
+    }
+
+    // @TODO find a better way to fight circular dependency.
```
[the remark this answers is not in the document]
Comment (cristianvasquez) [reply lib/View.js:54]: Yes, I don't know yet how to disentangle things like

 CubeSource.js > Source.js > View.js >  CubeSource.js
 
 Without breaking backward compatibility.

File: lib/Node.js (hunk 3 above)
Comment (cristianvasquez) [reply lib/Node.js]: You're right is as simple as that.
( I'm always in fear of terms not being equal :) )

**Pull request closed without merging**